### PR TITLE
Speedup (grouped)limitedCandidates by changing the implementation of the candidates' number limitation.

### PR DESCRIPTION
**Pull request by VinInn**
Currenlty, in both (grouped)limitedCandidates, for each "step" candidates are first all collected and then the size is reduced based on some score.
This is equivalent to not "pushing" a new candidate if worse than the current worst once the collection has reached the max allowed size. (but for candidates with equal score). 
The implementation has been therefore modified to this latter mechanism that is more performant.

It should also be noticed that the final sorting is not required and indeed in groupedLimitedCandidates there is NO final sort.

The Size of TempTrajectory has also been reduced to speed up move and swap.
Took the opportunity to reduce the size of Trajectory as well
and to cleanup  in general CkfTrajectoryBuilder code.

In a HLT menu
time of limitedCandidates improves of 20%
while the one of groupedLimitedCandidates improves of a bit more than 10%

HLT throughput improves of a solid 1.5% at least.

Purely technical.  Negligible regressions may appear in case of "candidates" with identical score.

### CODE REVIEW

**Comment by VinInn** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
> Took the opportunity to reduce the size of Trajectory as well.

is this from reordering the data members?

**Comment by slava77** [comment]
> Sorting of TempTrajectory in CkfTrajectoryBuilder::limitedCandidates takes almost 1% of HLT time.
> Reducing it's size to two pointers this timing is reduced by half.

wouldn't it be more practical to update `CkfTrajectoryBuilder::limitedCandidates` to sort indices to trajectories (or did I misunderstand the comment/code)?

What's the cost of now having more scattered data in memory? Is it negligible? It would be nice to see some profiler or at least the timing piechart.

**Comment by VinInn** [comment]
@slava77 , sorting indices and caching score. maybe yes. Was an option I considered. Not sure it is more practical. 
The data are not really much more scattered in memory, there is just one more (fast) indirection.

As said the code is now faster at least for HLT. In reco it crashes for reasons to be understood (not obvious at all: it seems that not everything has ben recompiled?)
will now try some relval.

**Comment by VinInn** [comment]
> > Took the opportunity to reduce the size of Trajectory as well.
> 
> is this from reordering the data members?

In part. also moving from short (int16) to uint8 .

**Comment by VinInn** [comment]
crash reason understood: comes from a test of TempTrajectory validity.
A clever fix did not work. Will try another or just revert to a full default constructor...

**Comment by VinInn** [comment]
@cmsbuild please test

**Comment by mmusich** [comment]
type performance-improvements

**Comment by VinInn** [comment]
strange as I run successfully in 11634.0_TTbar_14TeV+2021

**Comment by VinInn** [comment]
apparently the MULTIARCH (v3) was not crashing while the standard sse3 was....

**Comment by VinInn** [comment]
@cmsbuild please test

**Comment by VinInn** [comment]
I suppose 2500.2 is not caused by this PR

**Comment by VinInn** [comment]
@cmsbuild please test

**Comment by VinInn** [comment]
1) the failure does not seem to be associated to this PR
2) the DQM differences are expected. Maybe a high stat validation is required as the logic of limitedCandidates is in principle changed

**Comment by jfernan2** [comment]
assign hlt

**Comment by mmusich** [comment]
@cmsbuild, please test
    * failures are transient

**Comment by slava77** [comment]
> * [Reco comparison results: 2036 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_1_X_2024-05-07-1100+2dcb23/62699/validateJR.html)
> 
> * DQMHistoTests: [Total failures: 1532](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_1_X_2024-05-07-1100+2dcb23/62699/validateJR.html)

it looks like the differences are predominantly in gsf tracks, 
- some are in HLT displaced tracking 141.044 https://tinyurl.com/2c324rx8
- HLT GSF 140.046 https://tinyurl.com/28v7ud4r
- 136.874 [reco](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_1_X_2024-05-07-1100+2dcb23/62699/validateJR/136.874_RunEGamma2018C/all_reRECO_step3/) and [DQM](https://tinyurl.com/2asm7ns2); somewhat similar in [reco 136.793](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_1_X_2024-05-07-1100+2dcb23/62699/validateJR/136.793_RunDoubleEG2017C/all_reRECO_step3/)

at least with this 100 event x a few workflows stats, the differences seem small.

@VinInn what's the change in logic in the latest variant?

File: RecoTracker/CkfPattern/src/CkfTrajectoryBuilder.cc [hunk 1]
```diff
@@ -235,8 +139,19 @@ unsigned int CkfTrajectoryBuilder::limitedCandidates(const std::shared_ptr<const
           updateTrajectory(newTraj, std::move(*itm));
 
           if (toBeContinued(newTraj)) {
-            newCand.push_back(std::move(newTraj));
-            std::push_heap(newCand.begin(), newCand.end(), trajCandLess);
+            if (full) {
```
Comment (VinInn) [inline RecoTracker/CkfPattern/src/CkfTrajectoryBuilder.cc:142]: Here now we immediately check if new candiate is better of the worst in newCand.
if yes, we swap, if not we do not push it.
Of course we keep the "heap" properly ordered.

In principle the result should be identical to the previous one, unless some candidates have identical score as the order in not the same anymore...

I remind that this TrajectoryBuilder is used only by electron and muons....

File: RecoTracker/CkfPattern/src/CkfTrajectoryBuilder.cc [hunk 2]
```diff
@@ -199,14 +102,15 @@ unsigned int CkfTrajectoryBuilder::limitedCandidates(const std::shared_ptr<const
...
-  newCand.reserve(2 * theMaxCand);
+  newCand.reserve(theMaxCand);
 
   auto trajCandLess = [&](TempTrajectory const& a, TempTrajectory const& b) {
     return (a.chiSquared() + a.lostHits() * theLostHitPenalty) < (b.chiSquared() + b.lostHits() * theLostHitPenalty);
   };
 
   while (!candidates.empty()) {
     newCand.clear();
+    bool full = 0;
```
Comment (VinInn) [inline RecoTracker/CkfPattern/src/CkfTrajectoryBuilder.cc:113]: ops. should be "false", not 0. (of course 0 is false, so not a bug)

File: RecoTracker/CkfPattern/src/CkfTrajectoryBuilder.cc [hunk 3]
```diff
@@ -199,14 +102,15 @@ unsigned int CkfTrajectoryBuilder::limitedCandidates(const std::shared_ptr<const
   unsigned int nCands = 0;  // ignore startingTraj
   unsigned int prevNewCandSize = 0;
   TempTrajectoryContainer newCand;  // = TrajectoryContainer();
-  newCand.reserve(2 * theMaxCand);
+  newCand.reserve(theMaxCand);
```
Comment (VinInn) [inline RecoTracker/CkfPattern/src/CkfTrajectoryBuilder.cc:105]: Now the size cannot exceed _theMaxCand_

**Comment by VinInn** [comment]
added comments inline to further annotate the "new" implementation of the (same) logic to limit the number of candidate

**Comment by VinInn** [comment]
In principle the same logic (keep newCand size < MaxCand) can be also implemented in groupedLimimitedCandidates (actually advanceOneLayer)
I'm convinced that would be equivalent to the current one.
It would affect the whole tracking so if any iusse will be easy to spot in MTV

File: RecoTracker/CkfPattern/src/CkfTrajectoryBuilder.cc [hunk 4]
```diff
@@ -250,55 +165,9 @@ unsigned int CkfTrajectoryBuilder::limitedCandidates(const std::shared_ptr<const
...
-	newCand.resize(theMaxCand);
-      }
-      */
-
+      assert((int)newCand.size() <= theMaxCand);
+      if (full)
+        assert((int)newCand.size() == theMaxCand);
     }  // end loop on candidates
 
     std::sort_heap(newCand.begin(), newCand.end(), trajCandLess);
```
Comment (VinInn) [inline RecoTracker/CkfPattern/src/CkfTrajectoryBuilder.cc:173]: in groupedLimitedCandidates the candidates are not sorted, so we may remove this sorting as well?

**Comment by VinInn** [comment]
I added a printout and indeed there is plenty of candidates with the same score .
I wrote a small test of ```std::pop_heap```  and I confirm that the order of popping of equal elements depends on the way elements are pushed-back.

So, yes changing the logic will change the final candidate sample.
Is it relevant?  Not by construction. it is just an accident. 

I will implement the same logic in groupedLimitedCandidates as well.

**Comment by VinInn** [comment]
@cmsbuild please test

**Comment by VinInn** [comment]
I have run MTV on 1000 events from 11634.0  and I do not see any difference introduced by this PR
http://innocent.home.cern.ch/innocent/RelVal/ttbar141_NewTraj/

**Comment by VinInn** [comment]
here the results for 1000 events from 13034.0_TTbar_14TeV+2024PU
http://innocent.home.cern.ch/innocent/RelVal/ttbarPU141_NewTraj/
I do not see any hint to a non-negligible regression.

**Comment by VinInn** [comment]
For what I'm concerned this PR is final.
Does TRK-POG or Reco have more questions?
(BTW I do not see them in the list of people notified....)

**Comment by slava77** [comment]
> Does TRK-POG or Reco have more questions?

I looked at the 1K events tracking comparisons in https://github.com/cms-sw/cmssw/pull/44904#issuecomment-2105737242 and had no comments, sorry for not being explicit.

I'm not sure I've seen a profiling/timing result re my [comment from May 5](https://github.com/cms-sw/cmssw/pull/44904#issuecomment-2094836868)

**Comment by slava77** [comment]
>> I'm not sure I've seen a profiling/timing result

> I reported on time improvements in the (updated) first comment

thanks; apparently I was looking in time order and didn't check the PR description

**Comment by VinInn** [comment]
this is the perf report for the release

image


and this other one for this PR

image


the 20% improvement for limitedCandidate is pretty evident
for the Grouped version is a 5%

**Comment by slava77** [comment]
> the 20% improvement for limitedCandidate is pretty evident

Do I understand correctly that it is closer to 12-15%, if the overall `5-7%` decrease in unrelated modules like muonId or seedCreator are considered?

Is it fair to conclude also from looking at the unrelated modules that the grouped version is about the same if not slower?

**Comment by VinInn** [comment]
@slava77 I can try to run with more events to make initialization count less

**Comment by VinInn** [comment]
repeated the time measurements with 10K events and no GPU (one will notice now patatrack showing up)

This is the Release
image

and this the PR
image


I would say that the speed improvement in limitedCandidate is obvious and traceable to the reduction of "heap" operations.
In the grouped version the difference goes  "in the noise".

**Comment by slava77** [comment]
> I would say that the speed improvement in limitedCandidate is obvious and traceable to the reduction of "heap" operations.

+1

> In the grouped version the difference goes "in the noise".

both examples were worse. This time the unrelated modules are not visible to guess which direction the baseline is moving.
I don't see the `*_heap` operations in the baseline in the grouped, while it shows up at the bottom with the PR.


Is the test from HLT?
IIUC, the offline has a much larger weight in the grouped (even more so in a no-mkFit workflow/era). Perhaps a test there will be more convincing that the cost is at least not increasing.

**Comment by VinInn** [comment]
this is offline: first column Release, second colum PR
[innocent@patatrack02 13034.0_TTbar_14TeV+2024PU]$ paste TimeOriTraj.txt TimeNewTraj.txt | awk '{print $1, $2, $4}'
convTrackCandidates 0.207506 0.206641
conversionTrackCandidates 0.157866 0.144111
cosmicsVetoTrackCandidates 0.001399 0.001399
detachedQuadStepTrackCandidates 0.001174 0.001139
detachedQuadStepTrackCandidatesMkFit 0.012637 0.012584
detachedQuadStepTrackCandidatesMkFitSeeds 0.000278 0.000279
detachedTripletStepTrackCandidates 0.007312 0.007274
detachedTripletStepTrackCandidatesMkFit 0.340068 0.340722
detachedTripletStepTrackCandidatesMkFitSeeds 0.003492 0.003338
displacedRegionalStepTrackCandidates 0.181286 0.180114
duplicateDisplacedTrackCandidates 0.000020 0.000020
duplicateTrackCandidates 0.104824 0.105132
electronCkfTrackCandidates 0.011372 0.011065
highPtTripletStepTrackCandidates 0.035885 0.035877
highPtTripletStepTrackCandidatesMkFit 0.096393 0.096400
highPtTripletStepTrackCandidatesMkFitSeeds 0.001586 0.001528
initialStepTrackCandidates 0.006030 0.006038
initialStepTrackCandidatesMkFit 0.087755 0.087670
initialStepTrackCandidatesMkFitPreSplitting 0.092760 0.092999
initialStepTrackCandidatesMkFitSeeds 0.001295 0.001252
initialStepTrackCandidatesMkFitSeedsPreSplitting 0.001381 0.001337
initialStepTrackCandidatesPreSplitting 0.006307 0.006293
jetCoreRegionalStepBarrelTrackCandidates 0.054636 0.054550
jetCoreRegionalStepEndcapTrackCandidates 0.092552 0.091719
lowPtGsfEleCkfTrackCandidates 0.009014 0.008760
lowPtQuadStepTrackCandidates 0.704589 0.700201
lowPtTripletStepTrackCandidates 0.554903 0.552126
mixedTripletStepTrackCandidates 0.032825 0.032708
muonSeededTrackCandidatesInOut 0.003372 0.003375
muonSeededTrackCandidatesOutIn 0.000696 0.000695
muonSeededTrackCandidatesOutInDisplaced 0.000559 0.000560
pixelLessStepTrackCandidates 0.011386 0.010974
pixelLessStepTrackCandidatesMkFit 0.706225 0.707222
pixelLessStepTrackCandidatesMkFitSeeds 0.021475 0.020879
pixelPairStepTrackCandidates 0.152920 0.152784
tobTecStepTrackCandidates 0.502028 0.503019
uncleanedOnlyConversionTrackCandidates 0.002595 0.002350
uncleanedOnlyElectronCkfTrackCandidates 0.000094 0.000091

**Comment by VinInn** [comment]
Fro what concern unrelated modules:
  there are two for Pixels, plus all the functions below AdvanceOneLayer such as gropuedMeasurements

**Comment by slava77** [comment]
> this is offline: first column Release, second colum PR

totals after greping out `initialStep\|detached\|highPt\|pixelLess` and the rest would be more clear.

The results are pretty stable; it looks like the 3rd significant digit varies (things often vary more).

**Comment by VinInn** [comment]
I leave it as exercise to the reader...
Do you agree that is not worst?

**Comment by VinInn** [comment]
For sustainability, I think it  is better to keep the same implementation in both

**Comment by slava77** [comment]
> I leave it as exercise to the reader... Do you agree that is not worse

OK, only `convTrack\|displacedRegional\|jetCore\|lowPt\|mixed\|muon\|pixelPair\|tobTec` are using grouped; they add up to `old new new/old: 2.48787 2.47849 0.99623` so, around 0.4% better.
The mkfit reference ratio is 1.00026.

> For sustainability, I think it is better to keep the same implementation in both

fine

thanks for the inputs.
+1

**Comment by mmusich** [comment]
+hlt
   * based on https://github.com/cms-sw/cmssw/pull/44904#issuecomment-2117793724,
   * not clear though why the HLT signature was requested here.

**Comment by jfernan2** [comment]
+1

**Comment by rappoccio** [comment]
+1

**Pull request merged**